File: modules/google-earth-engine/docker/sepal-ee/sepal/rx/workqueue.py
import logging

from rx import Observable, empty, of
from rx.operators import catch, concat, delay, do_action, finally_action, flat_map, group_by, map, merge, merge_all, \
    observe_on, take_until, take_while
from rx.scheduler.threadpoolscheduler import ThreadPoolScheduler
from rx.subject import Subject
from sepal.rx.operators import take_until_disposed, throw_when
from sepal.rx.retry import retry_with_backoff


class WorkQueue(object):
    def __init__(self, concurrency_per_group, delay_seconds=0, description=None):
        self._queue = Subject()
        self._description = description
        self.request_scheduler = ThreadPoolScheduler(concurrency_per_group)
        producer_scheduler = ThreadPoolScheduler(concurrency_per_group)

        def on_next(result):
            output = result['output']
            output.on_next({'value': result.get('value'), 'completed': result.get('completed')})

        self._subscription = self._queue.pipe(
            observe_on(producer_scheduler),
            group_by(lambda r: r['group']),
            flat_map(
                lambda concurrency_group: concurrency_group.pipe(
                    map(lambda r: r['work']),
                    delay(delay_seconds),
                    merge(max_concurrent=concurrency_per_group),
                    merge_all(),
                )
            ),
            take_until_disposed()
        ).subscribe(
            on_next=on_next,
            on_error=lambda error: logging.exception('Error in {} request stream'.format(self)),
            scheduler=producer_scheduler
        )

    def enqueue(
            self,
            observable: Observable,
            group: str = 'default-group',
            retries: int = 0,
            description: str = None
    ) -> Observable:
        def log_status(status):
            logging.debug(str({
                'WorkQueue': str(self),
                'group': group,
                status: description
            }))

        log_status('ENQUEUED')
        output = Subject()
        errors = Subject()
        output_finalized = Subject()

        def handle_error(e, _):
            log_status('FAILED')
            errors.on_next(e)
            return empty()

        def set_output_finalized():
            output_finalized.on_next(True)

        work = of(True).pipe(
            do_action(lambda _: log_status('STARTED')),
            take_until(output_finalized),
            flat_map(
                lambda _: observable.pipe(
                    map(lambda value: of({'value': value, 'output': output})),
                    retry_with_backoff(
                        retries=retries,
                        description='{}.enqueue(group={}, description={})'.format(self, group, description)
                    ),
                    catch(handler=handle_error),
                    take_until(output_finalized),
                    take_until_disposed()
                )
            ),
            concat(of(of({'completed': True, 'output': output}))),
            finally_action(lambda: log_status('COMPLETED'))
        )

        self._queue.on_next({'work': work, 'group': group})

        return output.pipe(
            observe_on(self.request_scheduler),
            throw_when(errors),
            take_while(lambda r: not r.get('completed')),
            map(lambda r: r.get('value')),
            finally_action(set_output_finalized)
        )

    def dispose(self):
        if self._subscription:
            self._subscription.dispose()

    def __str__(self):
        return 'WorkQueue({})'.format(self._description) if self._description else super().__str__()

File: modules/gui/frontend/src/_/pytest/workqueue.js
import {Subject, isObservable, of} from 'rxjs'
import {catchError, concat, delay, filter, map, merge, mergeMap, share, takeUntil, takeWhile, tap} from 'rxjs/operators'
import {onUnsubscribe} from './operators'

export class WorkQueue {
    constructor({concurrency, delaySeconds = 0, description}) {
        this.requests$ = new Subject()
        this.outputSubject$ = new Subject()
        this.output$ = this.outputSubject$.pipe(share())
        this.description = description
        this.subscription = this.requests$.pipe(
            map(r => r.request),
            merge(concurrency),
            delay(delaySeconds)
        ).subscribe(
            request => this.outputSubject$.next(request),
            error => console.error(`Error in request stream: ${error}`),
            () => console.info('********** COMPLETED **********')
        )
    }

    enqueue({observable$, group = 'some-default-group', description}) {
        console.info(`Enqueue ${description}`)
        const key = Math.random()
        const canceled = new Subject()
        let error = undefined

        const logStatus = status => console.info({group, key, status})

        const handleError = e => {
            logStatus('FAILED')
            error = e
            return of({key, error})
        }

        const throwIfError = request => {
            if (error) {
                throw Error(error)
            } else {
                return of(request)
            }
        }

        const extractValue = value =>
            isObservable(value)
                ? value
                : of(value)

        const cancelRequest = () => {
            canceled.next(true)
            console.info(`*** Cancelling ${description}`)
        }

        logStatus('ENQUEUED')

        const request = of(true).pipe(
            tap(() => logStatus('STARTED')),
            mergeMap(() =>
                observable$.pipe(
                    mergeMap(extractValue),
                    tap(value => console.info(`Inner value: ${value}`)),
                    map(value => ({key, value})),
                    // retry with backoff
                    catchError(e => handleError(e)),
                    takeUntil(canceled)
                )
            ),
            concat(of({key, complete: true}).pipe(
                tap(() => logStatus('COMPLETED'))
            ))
        )

        this.requests$.next({request})

        return this.output$.pipe(
            filter(request => request.key === key),
            mergeMap(request => throwIfError(request)),
            takeWhile(request => !request.complete),
            mergeMap(request => of(request.value)),
            onUnsubscribe(cancelRequest)
        )
    }

    dispose() {
        this.subscription && this.subscription.unsubscribe()
    }
}
